plugin/trino-loki/src/main/java/io/trino/plugin/loki/LokiMetadata.java

```diff
@@ -0,0 +1,120 @@
+import static io.trino.plugin.loki.LokiTableFunction.SCHEMA_NAME;
+import static io.trino.spi.type.DoubleType.DOUBLE;
+import static io.trino.spi.type.TimestampWithTimeZoneType.TIMESTAMP_TZ_NANOS;
+import static io.trino.spi.type.TypeSignature.mapType;
+import static io.trino.spi.type.VarcharType.VARCHAR;
+import static java.util.Objects.requireNonNull;
+
+public class LokiMetadata
+        implements ConnectorMetadata
+{
+    private static final String TABLE_NAME = "default";
+
+    private final Type labelsMapType;
+
+    private final LokiClient lokiClient;
+
+    @Inject
+    public LokiMetadata(LokiClient lokiClient, TypeManager typeManager)
+    {
+        this.lokiClient = requireNonNull(lokiClient, "lokiClient is null");
+        labelsMapType = typeManager.getType(mapType(VARCHAR.getTypeSignature(), VARCHAR.getTypeSignature()));
+    }
+
+    @Override
+    public Optional<TableFunctionApplicationResult<ConnectorTableHandle>> applyTableFunction(ConnectorSession session, ConnectorTableFunctionHandle handle)
+    {
+        if (!(handle instanceof LokiTableFunction.QueryHandle queryHandle)) {
+            return Optional.empty();
+        }
+
+        LokiTableHandle tableHandle = queryHandle.getTableHandle();
+        return Optional.of(new TableFunctionApplicationResult<>(tableHandle, tableHandle.columnHandles()));
+    }
+
+    @Override
+    public ConnectorTableMetadata getTableMetadata(ConnectorSession session, ConnectorTableHandle table)
+    {
+        LokiTableHandle lokiTableHandle = (LokiTableHandle) table;
+
+        List<ColumnMetadata> columns = lokiTableHandle.columnHandles()
+                .stream()
+                .map(LokiColumnHandle.class::cast)
+                .map(LokiColumnHandle::columnMetadata)
+                .collect(toImmutableList());
+
+        return new ConnectorTableMetadata(new SchemaTableName(SCHEMA_NAME, TABLE_NAME), columns);
+    }
+
+    @Override
+    public ConnectorTableHandle getTableHandle(
+            ConnectorSession session,
+            SchemaTableName tableName,
+            Optional<ConnectorTableVersion> startVersion,
+            Optional<ConnectorTableVersion> endVersion)
+    {
+        throw new TrinoException(LokiErrorCode.LOKI_TABLE_ERROR, "Loki connector does not support querying tables directly. Use the TABLE function instead.");
+    }
+
+    @Override
+    public ColumnMetadata getColumnMetadata(ConnectorSession session, ConnectorTableHandle tableHandle, ColumnHandle columnHandle)
+    {
+        LokiColumnHandle lokiColumnHandler = (LokiColumnHandle) columnHandle;
+        return lokiColumnHandler.columnMetadata();
+    }
+
+    public List<ColumnHandle> getColumnHandles(String query)
+    {
+        ImmutableList.Builder<ColumnHandle> columnsBuilder = ImmutableList.builderWithExpectedSize(3);
+        columnsBuilder.add(new LokiColumnHandle("labels", this.labelsMapType, 0));
+        columnsBuilder.add(new LokiColumnHandle("timestamp", TIMESTAMP_TZ_NANOS, 1));
```

Comment:
Since we return the timestamp with time zone as long i.e we need to expose them as `TIMESTAMP_TZ_MILLIS`